rfcs/0000-error-handling.md

```diff
@@ -0,0 +1,144 @@
+# RFC-0000: Subgraph error handling and reporting
+
+<dl>
+  <dt>Author</dt>
+  <dd>Leonardo Y. Schwarzstein</dd>
+    <dt>RFC pull request</dt>
+  <dd>URL</dd>
+
+  <dt>Date of submission</dt>
+  <dd>2020-03-19</dd>
+
+  <dt>Date of approval</dt>
+  <dd>YYYY-MM-DD</dd>
+
+  <dt>Approved by</dt>
+  <dd>First Person, Second Person</dd>
+</dl>
+
+## Contents
+
+<!-- toc -->
+
+## Summary
+
+Synced subgraphs no longer stop syncing on errors. Failed subgraphs return an error along with the data when queried.
+
+The indexing status API is improved and documented.
+
+## Goals & Motivation
+
+Currently, an error in a WASM trigger handler will cause a subgraph to be marked as failed and stop syncing. The proposal is for it still be marked as failed, but discard any changes made by the failed handler and attempt to continue. In the cases where the failure is catastrophic and all handlers start failing, this will not do much, but in the cases where the subgraph can continue operational with a bit of missing data, this will allow applications to continue running on up-to-date data and give developers more time to fix their subgraph.
+
+Less catastrophic failures do have a downside: It's harder to notice that anything is wrong at all. We'll include an error in the graphql response for failed subgraphs. We will improve and document the indexing status API, so that developers can programmatically check details on a subgraph's health and build their own alerts and dashboards.
+
+## Urgency
+
+Sooner is better than later.
+
+## Terminology
+
+No new terminology, but the semantics of the "Failed" status are altered.
+
+## Detailed Design
+
+### Error handling
+
+If a WASM handler terminates with a trap, that is an error. The node discards all entity operations and data source operations done by the handler and continue syncing.
+
+For subgraphs that are not yet synced, any handler error halts syncing, so errors in pending versions will not make their way to the current version. 
+
+When the node starts, we currently reset the failed flag on subgraphs. The flag is no longer be reset on synced subgraphs, since the subgraph will already be past the errors.
+
+Any graphql queries to a failed subgraph will include an error of the form:
+
+```
+{
+	message: "Subgraph failed"
```

Comment:
Can we just use the `Health` status here? I'd prefer a machine-readable enum over essentially freeform text.

Reply:
As in "Subgraph healthy/unhealthy/failed" or just "healthy/unhealthy/failed"?

Reply:
Just "healthy/unhealthy/failed" - something that is guaranteed to stay that way, and whose exact text is part of the API so that people can reliably check for it.

Reply:
Should we use all caps `"UNHEALTHY"` so it looks more like a constant?

Reply:
I'd say lowercase is fine.

Reply:
I've updated this.